### DISCUSSION TITLE
AVR32DA28 analogWrite to PA0 doesn't work by default

### DISCUSSION THREAD

**Discussion by grandaspanna** [SpenceKonde/DxCore#350] [Q&A] [answered]
I just wanted to check my understanding. I'm working through a new board design with an AVR32DA28 and wasn't getting any PWM on PA0. This code generates nothing:
```
void setup() {
  // put your setup code here, to run once:

  pinMode(PIN_PA0,OUTPUT);
  
}

void loop() {
  // put your main code here, to run repeatedly:

  analogWrite(PIN_PA0,127);
}
```

If I replace the analogWrite with a digitalWrite to HIGH, I get a voltage on PA0.

If I run this code, I get a PWM signal:
```
void setup() {
  // put your setup code here, to run once:

  pinMode(PIN_PA0,OUTPUT);
  PORTMUX.TCAROUTEA   = (PORTMUX.TCAROUTEA & ~(PORTMUX_TCA0_gm)) | PORTMUX_TCA0_PORTA_gc;
}

void loop() {
  // put your main code here, to run repeatedly:

  analogWrite(PIN_PA0,127);
}
```

Is this expected?

**Comment by SpenceKonde** [comment, marked as answer]
Yes, I thought uit was clearly marked on the charts, but after my butchering of them it may not be as clear

The PWM code for TCA's on a Dx or future Ex part detect the state of TCAROUTEA when called, to find out if they can PWM that pin,. 
During initialization, we set the initial mux to somewhere else in most cases - PORTA is the last place you want your PWM pins to be on a DA/DB! What with 2 pins taken up potentially with a crystal, and 2 more as possible mux options for TCD (and they never fixed the silicon bug that breaks the other mux options for TCD) . In DD you will be able to do the same thing with the TCDs. However, we put them on the "most otherwise useless" set of pins otherwise, so the pinout chart can show default pin and users who are new can just use them, but if you need other peripherals, you can shove the timers around as needed. The only ones you can't are the TCBs,. The TCBs, because they move individually, have higher overhead to do, and sincethey are a shitty way to get PWM, the reward for that is small

Note that as described in docs it is highly recommended to not have any PWM running either on the timer, or on the pins that you are moving the timer to), In the first case, they'll stay active, (but if the pin is set input you won't know,- but a pinMode that sets it output would result in surprise PWM there. (only digitalWrite, not pinMode invokes turnOnffPWM, because that is a very expensive function call. 
```
uint8_t next=0
PORTMUX.TCAROUTEA=0;
uint8_t mypin = PIN_PA5; // This makes a very dramatic example because there are TWO movable timers on it.
pinMode(mypin,OUTPUT); - hard to gauge the speed with the pin set input ;-)
while(1) {
  digitalWrite(PIN_PA5, next);  // Change the output so you can see what the frequency is by attaching a scope probe. 
  if (next==1) {
    next = 0;
  } else {
    next = 1;
  }
}
```
look at the frequency of the squarewave you get out of that. Compare it to a pin that has no PWM output at all. (I think PD7 has no PWM out?) - it should be markedly faster. 

Then make mypin volatile, so it can't assume the value even though it's never changed
- this should be slower on all pins, 

Finally, try it with PORTMUX set to something else. 

Compare to PIN_PD7. 

This shows what happens when youstrip the pinwrites down to the bare bones. 
```
uint8_t next=0
PORTMUX.TCAROUTEA=0;
//uint8_t mypin = PIN_PA5; // same pin - will probably be the same with this line, but here I wanted to make sure the code did what I said
pinMode(mypin,OUTPUT); 
while(1) { //
  digitalWriteFast(PIN_PA5, next);  // Change the output so you can see what the frequency is by attaching a scope probe. 
  if (next==1) {
    next = 0;
  } else {
    next = 1;
  }
}
```
digitalWriteFast will now be like 2-3 clock cycles (when passed a non-constant second argument, there will be a couple of extra clocks. 
I think the optimizer can see that next is either 0 or 1, and hence won't have to check for CHANGE, there's still gotta be a few instructions to know whether to execute cbi 1, 5 or sbi 1,5). digitalWriteFast() with both arguments compiles to a single instruction .
digitalWrite can execute uh.... a lot more.

In the first example, if you put analogWrite(PIN_PA5,127); delay(1000); after setting the PORTMUX,.  the PWM would show for 1 second. In the secondexample, doing the same thing, the PWM would never stop, because fast digital IO doesn't turn off PWM. Otherwise we would have to call it digitalWriteSlow, and we already have digitalWrite().

**Reply by grandaspanna** [reply]
Thanks for the explanation. My bad for not reading the pinout chart carefully, I just went with the default WO0.. from the datasheet and didn't think too much about it.
